haskell/test/SMT/Model/Proposal.hs

```diff
@@ -0,0 +1,416 @@
+  pure (frozenTotalSupply + tokens)
+
+unlockGovernanceTokens :: Natural -> Address -> ModelT Natural
+unlockGovernanceTokens tokens addr = do
+  frozenTotalSupply <- getStore <&> sFrozenTotalSupply
+  governanceToken <- getStore <&> sGovernanceToken
+  selfAddr <- get <&> msSelfAddress
+  let param = FA2.TransferItem
+            { tiFrom = selfAddr
+            , tiTxs = [ FA2.TransferDestination
+                { tdTo = addr
+                , tdTokenId = (governanceToken & gtTokenId)
+                , tdAmount = tokens
+                } ]
+            }
+  _ <- makeTransferOnToken [param] (governanceToken & gtAddress)
+  case frozenTotalSupply `minusNaturalMaybe` tokens of
+    Just n -> pure n
+    Nothing -> throwError BAD_STATE
+
+
+applyUnfreeze :: BaseDaoContractCall -> UnfreezeParam -> ModelT ()
+applyUnfreeze cc param = do
+  let senderAddr = cc & ccSender
+  store <- getStore
+  let amt = arg #amount param
+  currentStage <- getCurrentStageNum
+
+  newFh <- case Map.lookup senderAddr (store & sFreezeHistory & T.unBigMap) of
+        Just fh -> do
+          updatedFh <- fh
+              & updateFh currentStage
+              & subFrozenFh amt
+          pure $ BigMap $ Map.insert senderAddr updatedFh (store & sFreezeHistory & T.unBigMap)
+        Nothing -> throwError NOT_ENOUGH_FROZEN_TOKENS
+
+  frozenTotalSupply <- unlockGovernanceTokens amt senderAddr
+
+  modifyStore $ \s -> s { sFreezeHistory = newFh, sFrozenTotalSupply = frozenTotalSupply}
+
+isLevelReached :: Proposal -> Natural -> ModelT Bool
+isLevelReached proposal target = do
+  lvl <- get <&> msLevel
+  pure (lvl > (proposal & plStartLevel) + target)
+
+-- Note: Currently, resulted operation from proposal is not taking into account.
+handleProposalIsOver :: Natural -> ProposalKey -> Counter -> ModelT Counter
+handleProposalIsOver proposalLvl proposalKey counter = do
+  -- store <- getStore
+  config <- getConfig
+  proposal <- checkIfProposalExist proposalKey
+
+  isExpired <- isLevelReached proposal =<< (getConfig <&> cProposalExpiredLevel)
+  canBeFlushed <- isLevelReached proposal =<< (getConfig <&> cProposalFlushLevel)
+  if isExpired then
+    throwError EXPIRED_PROPOSAL
+  else if canBeFlushed && (counter & cCurrent) < (counter & cTotal) then do
+    let counter_ = counter { cCurrent = (counter & cCurrent) + 1}
+    cond <- getStore <&> \s ->
+              doTotalVoteMeetQuorumThreshold proposal s && ((proposal & plUpvotes) > (proposal & plDownvotes))
+
+    unfreezeProposerAndVoterToken cond proposal
+    case cond of
+      True -> do
+        (_, newExtra) <- getStore >>= \s -> ((config & cDecisionLambda) # pair) -$$? (proposal, (s & sExtra))
+        modifyStore $ \s -> s { sExtra = newExtra }
+      False -> pure ()
+    -- skip operations
+    deleteProposal proposalLvl proposalKey
+
+    pure counter_
+  else
+    pure counter
+
+-- | Counter type used in `flush`
+data Counter = Counter {cCurrent :: Natural, cTotal :: Natural}
+  deriving stock (Show)
```

Comment:
This too. Do we actually need a counter?

AFAIK we have one in LIGO because that's the only way to keep track of how many (valid and complete) proposals we have flush.

In Haskell however we can `filter` and then `take` the amount that we want and that should be enough.

Please rewrite the `flush` entrypoint as well.

Reply:
Yeah, we don't actually need counter. I've also updated the `flush` entrypoints.